File: banking-ui/src/utils/money.js
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/** Canonical Money utilities. No legacy field normalization or FX. */
export const currencyExponents = Object.freeze({ USD: 2, MXN: 2, JPY: 0, BHD: 3 });

export function validateMoney(money) {
  if (!money || !Number.isSafeInteger(money.amount_minor) ||
      !Object.hasOwn(currencyExponents, money.currency_code)) {
    throw new Error('Invalid Money amount or currency.');
  }
  return money;
}

export function parseMoneyInput(input, currencyCode) {
  if (!Object.hasOwn(currencyExponents, currencyCode)) throw new Error('Unsupported currency.');
  const exponent = currencyExponents[currencyCode];
  const match = /^([+-]?)([0-9]+)(?:\.([0-9]+))?$/.exec(input.trim());
  if (!match || (match[3] || '').length > exponent) {
    throw new Error(`Enter an ungrouped amount with at most ${exponent} decimal places, using a decimal point.`);
  }
  const minor = BigInt(match[2]) * (10n ** BigInt(exponent)) + BigInt((match[3] || '').padEnd(exponent, '0') || '0');
  const signed = match[1] === '-' ? -minor : minor;
  if (signed > BigInt(Number.MAX_SAFE_INTEGER) || signed < -BigInt(Number.MAX_SAFE_INTEGER)) {
    throw new Error('Amount exceeds the supported range.');
  }
  return { amount_minor: Number(signed), currency_code: currencyCode };
}

export function formatMoney(money, locale = 'en-US') {
  if (money == null) return '—';
  validateMoney(money);
  const exponent = currencyExponents[money.currency_code];
  const value = BigInt(money.amount_minor);
  const digits = (value < 0n ? -value : value).toString().padStart(exponent + 1, '0');
  const decimal = (value < 0n ? '-' : '') + (exponent
    ? `${digits.slice(0, -exponent)}.${digits.slice(-exponent)}` : digits);
  // Intl accepts an exact decimal string, avoiding Number division/rounding.
  return new Intl.NumberFormat(locale, { style: 'currency', currency: money.currency_code,
    minimumFractionDigits: exponent, maximumFractionDigits: exponent }).format(decimal);
}

export function sumMoneyByCurrency(values) {
  const totals = new Map();
  for (const value of values) {
    validateMoney(value);
    totals.set(value.currency_code, (totals.get(value.currency_code) || 0n) + BigInt(value.amount_minor));
  }
  return [...totals].map(([currency_code, amount]) => validateMoney({currency_code, amount_minor: Number(amount)}));
}

export function paymentIntent(previous, request, createKey = () => crypto.randomUUID()) {
  const signature = JSON.stringify([request.source_account_id, request.credit_account_id,
    request.money.currency_code, request.money.amount_minor]);
  return previous?.signature === signature ? previous : { signature, key: createKey(), completed: false };
}

File: banking-service/models/money.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Canonical integer Money with a bounded historical USD reader."""

from types import MappingProxyType
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StrictInt

# The common range supported losslessly by PostgreSQL BIGINT and JS numbers.
MAX_MINOR = 2**53 - 1
CurrencyCode = Literal["USD", "MXN", "JPY", "BHD"]
CURRENCY_EXPONENTS = MappingProxyType({"USD": 2, "MXN": 2, "JPY": 0, "BHD": 3})
MinorAmount = Annotated[StrictInt, Field(ge=-MAX_MINOR, le=MAX_MINOR)]


class Money(BaseModel):
    """Signed value; journal entry amounts are positive with separate direction."""

    model_config = ConfigDict(
        frozen=True,
        extra="forbid",
        json_schema_extra={"examples": [
            {"amount_minor": 149900, "currency_code": "USD"},
            {"amount_minor": 125, "currency_code": "JPY"},
            {"amount_minor": 1234, "currency_code": "BHD"},
        ]},
    )
    amount_minor: MinorAmount
    currency_code: CurrencyCode

    @property
    def exponent(self) -> int:
        return CURRENCY_EXPONENTS[self.currency_code]


def from_legacy_usd(amount_cents: int) -> Money:
    """Read immutable historical USD values without numeric coercion."""
    return Money(amount_minor=amount_cents, currency_code="USD")


def money_fields(name: str, money: Money) -> dict:
    """Serialize only the canonical named Money field."""
    return {name: money.model_dump(mode="json")}
